recipes/ginkgo/all/conanfile.py

```diff
@@ -0,0 +1,100 @@
+                                         "7.4", "7.5", "8", "8.1", "8.2", "8.3",
+                                         "8.4", "9", "9.1", "9.2", "9.3", "10",
+                                         "10.1"]},
+                             "clang": {"runtime": None, "libcxx": None,
+                                       "version": [
+                                           "3.9", "4.0", "5.0", "6.0", "7.0",
+                                           "7.1", "8", "9", "10", "11"]},
+                             "apple-clang": {"libcxx": None, "version": [
+                                                 "10.0", "11.0", "12.0"]},
+                             "intel": {"version": ["18", "19"]}},
+                "build_type": None, "arch": None}
+    options = {"shared": [True, False], "fPIC": [True, False], "openmp": [
+        True, False], "cuda": [True, False]}
+    default_options = {"shared": False, "fPIC": False,
+                       "openmp": False, "cuda": False}
+    generators = "cmake"
+    exports_sources = ["CMakeLists.txt", "patches/**"]
+
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename(self.name + "-" + self.version, self._source_subfolder)
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.definitions["GINKGO_BUILD_TESTS"] = False
+        self._cmake.definitions["GINKGO_BUILD_EXAMPLES"] = False
+        self._cmake.definitions["GINKGO_BUILD_BENCHMARKS"] = False
+        self._cmake.definitions["GINKGO_DEVEL_TOOLS"] = False
+        self._cmake.definitions["GINKGO_BUILD_REFERENCE"] = True
+        self._cmake.definitions["GINKGO_BUILD_OMP"] = self.options.openmp
+        self._cmake.definitions["GINKGO_BUILD_CUDA"] = self.options.cuda
+        self._cmake.definitions["GINKGO_BUILD_HIP"] = False
+        self._cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
+        if "fPIC" in self.options:
+            self._cmake.definitions["CMAKE_POSITION_INDEPENDENT_CODE"] = self.options.fPIC
+        self._cmake.configure(build_folder=self._build_subfolder)
+        return self._cmake
+
+    def build(self):
+        for patch in self.conan_data["patches"][self.version]:
+            tools.patch(**patch)
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
```

Comment:
Judging by the size of this project this might not be easy....

Please look into adding support for CMake components to model https://github.com/ginkgo-project/ginkgo/blob/develop/cmake/GinkgoConfig.cmake.in

At a minimum it would be appreciate to have some like this https://github.com/conan-io/conan-center-index/blob/0695c9d81f94a9a2ee9ecf25b82dab7a19ea38eb/recipes/span-lite/all/conanfile.py#L31

Reply:
I'm not sure I completely understand what you mean. I fear mirroring the GinkgoConfig.cmake file in the conanfile is not really an option, since we would like to avoid having to maintain multiple descriptions of the same setup, and in its current setting, it only exports CMake variables describing the environment Ginkgo was built in anyways.

Ginkgo doesn't really have components per se, it generates a number of different libraries, which can only be used if linked together. This is the reason I needed to model their dependencies in package_info(self).

Reply:
Hmm, I did not look too heavily once I saw how big it was! but checking into this AM I have a few suggestions

> in its current setting, it only exports CMake variables describing the environment

Yep, Conan does all that already so I agree with position we should not do that!

However, there's _one_ exported target that I can find which is `Ginkgo::Ginkgo` 

https://github.com/ginkgo-project/ginkgo/blob/e0e0cc47dfe0cf48ca74198a488ab8d07a4f8719/cmake/install_helpers.cmake#L53
https://github.com/ginkgo-project/ginkgo/blob/e0e0cc47dfe0cf48ca74198a488ab8d07a4f8719/cmake/install_helpers.cmake#L91

however this currently is not generated by Conan.

Please add these two lines to your `package_info` this was consumer can have the same interface as if they installed from source with `cmake --install .`

```py
self.cpp_info.names["cmake_find_package"] = "Ginkgo"
self.cpp_info.names["cmake_find_package_multi"] = "Ginkgo"
```